src/eval.c

```diff
@@ -3665,6 +3674,367 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+    to_eval_current = (char_u *) current.ga_data;
+    if (!eval1(&to_eval_current, &current_result, evaluate))
+    {
+	ga_clear(&current);
+	ga_clear(&result);
+	clear_tv(&current_result);
+	return FAIL;
+    }
+    ga_concat_free(&result, current_result.vval.v_string);
+
+    /*
+     * Return the result.
+     */
+    rettv->v_type = VAR_STRING;
+    rettv->v_lock = 0;
+    rettv->vval.v_string = (char_u *) result.ga_data;
+
+    ga_clear(&current);
+    return OK;
+}
+
+/*
+ * Refreshes previous_quote with NUL
+ *	if the closing quote of
+ *	the opening quote (previous_quote) found on **source.
+ * Refreshes previous_quote with ' or "
+ *	if a opening quote is never found and that quote found.
+ * Doesn't Refresh if no quotes found.
+ */
+    static void
+refresh_pair_of_quote(char_u *previous_quote, char_u source)
+{
+    if (*previous_quote == NUL && source == '\'')
+    {
+	*previous_quote = '\'';
+	return;
+    }
+    else if (*previous_quote == NUL && source == '"')
+    {
+	*previous_quote = '"';
+	return;
+    }
+
+    if (*previous_quote == '\'' && source == '\'')
+	*previous_quote = NUL;
+    else if (*previous_quote == '"' && source == '"')
+	*previous_quote = NUL;
+}
+
+    static int
+is_escaped_quote(int is_literal_string, char_u *p)
+{
+    return
+	(is_literal_string && *p == '\'' && *(p + 1) == '\'') ||
+	(!is_literal_string && *p == '\\' && *(p + 1) == '"');
+}
+
+/*
+ * Finds closing '}' of taken expr
+ */
+    static int
+forward_to_end_of_template_expr(char_u **expr, int is_literal_string)
+{
+    char_u  *expr_head = *expr;
+    size_t  nested_blocks = 0;
+    char_u  encountered_quote = NUL;
+
+    // While the closing '}' encountered. The '}' is neither '}' of a dict nor
+    // '}' in a string.
+    for (; !(encountered_quote == NUL && nested_blocks == 0 && **expr == '}');
+							MB_PTR_ADV(*expr))
+    {
+	switch (**expr)
+	{
+	    case '{':
+		if (encountered_quote == NUL)
+		    ++nested_blocks;
+		break;
+	    case '}':
+		if (encountered_quote == NUL)
+		    --nested_blocks;
+		break;
+	    case NUL:
+		semsg(_("E451: Unterminated template expression: %s"), expr_head);
+		return FAIL;
+	    default:
+		if (is_escaped_quote(encountered_quote == '\'', *expr))
+		{
+		    ++*expr;
+		    break;
+		}
+		refresh_pair_of_quote(&encountered_quote, **expr);
+		break;
+	}
+    }
+
+    if (*expr == expr_head)
+    {
+	semsg(_("E452: Empty template expression: %s"), expr_head);
+	return FAIL;
+    }
+
+    return OK;
+}
+
+/*
+ * Get expr of "${" expr  "}" of source,
+ * and forward source.
+ */
+    static char_u*
+read_template_expr(char_u **source, int is_literal_string)
+{
+    char_u  *expr_head;
+    char_u  *expr;
+
+    *source += 2;  // forward with beginning '${'
+    expr_head = *source;
+
+    if (!forward_to_end_of_template_expr(source, is_literal_string))
+	return NULL;
+
+    **source = NUL;
+    expr = (char_u *) alloc(STRLEN(expr_head) + 1);
+    STRCPY(expr, expr_head);
+    **source = '}';
```

Comment:
These 4 lines can be simpler:

```C
    expr = vim_strnsave(expr_head, (int)(*source - expr_head));
```

Or

```C
    return vim_strnsave(expr_head, (int)(*source - expr_head));
```

and remove the `expr` variable from this function.

Reply:
I'll use `    return vim_strnsave(expr_head, (int)(*source - expr_head));` !